File: ytmp3.js
#!/usr/bin/env node
/**
 * ytMP3 - YouTube to MP3 Downloader (Node.js version)
 * A simple CLI tool to download YouTube playlists as MP3 files
 */

const { execSync } = require('child_process');
const readline = require('readline');
const fs = require('fs');
const path = require('path');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

function question(query) {
  return new Promise((resolve) => rl.question(query, resolve));
}

function getDefaultOutputDir() {
  const now = new Date();
  const year = now.getFullYear();
  const month = String(now.getMonth() + 1).padStart(2, '0');
  const day = String(now.getDate()).padStart(2, '0');
  return `${year}-${month}-${day}-ytmp3`;
}

async function checkDependency(cmd) {
  try {
    if (cmd === 'ffmpeg') {
      execSync(cmd, { stdio: 'pipe', timeout: 5000 });
    } else {
      execSync(`${cmd} --version`, { stdio: 'pipe' });
    }
    return true;
  } catch (error) {
    // Even if it fails with "Unrecognized option", if we got output, it's working
    if (error.stderr && error.stderr.toString().includes('ffmpeg version')) {
      return true;
    }
    return false;
  }
}

function downloadAsMp3(url, outputPath) {
  return new Promise((resolve, reject) => {
    // Create output directory if it doesn't exist
    if (!fs.existsSync(outputPath)) {
      fs.mkdirSync(outputPath, { recursive: true });
    }

    const ytdlpCmd = `yt-dlp --extract-audio --audio-format mp3 --audio-quality 192k -o "${path.join(outputPath, '%(title)s.%(ext)s')}" "${url}"`;

    console.log(`\nStarting download...\nCommand: ${ytdlpCmd}\n`);

    try {
      const child = execSync(ytdlpCmd, {
        stdio: 'pipe',
        encoding: 'utf8',
      });

      console.log(child.stdout);
      if (child.stderr) {
        console.error('Stderr:', child.stderr);
      }

      console.log(`\nDownload completed! Files saved to: ${outputPath}`);
      resolve();
    } catch (error) {
      reject(new Error(`Download failed: ${error.message}`));
    }
  });
}

async function main() {
  console.log('=== ytMP3 - YouTube to MP3 Downloader (Node.js version) ===\n');

  try {
    // Check dependencies
    const ffmpegAvailable = await checkDependency('ffmpeg');
    const ytdlpAvailable = await checkDependency('yt-dlp');

    if (!ffmpegAvailable) {
      console.error('Error: FFmpeg is not installed or not in PATH');
      console.log('Please install FFmpeg: https://ffmpeg.org/download.html');
      return 1;
    }

    if (!ytdlpAvailable) {
      console.error('Error: yt-dlp is not installed or not in PATH');
      console.log(
        'Please install yt-dlp: https://github.com/yt-dlp/yt-dlp#installation'
      );
      return 1;
    }

    // Get YouTube URL
    let url = '';
    while (!url.trim()) {
      url = await question('Enter the YouTube video or playlist URL: ');
      if (!url.trim()) {
        console.log('Error: URL cannot be empty\n');
      }
    }
    url = url.trim();

    // Get custom output directory
    const defaultDir = getDefaultOutputDir();
    let customDir = await question(
      `Enter download location (default: ${defaultDir}): `
    );
    customDir = customDir.trim();

    const outputPath = customDir || defaultDir;

    // Confirm before starting
    await question('\nPress Enter to start downloading...');

    // Start download
    await downloadAsMp3(url, outputPath);

    console.log('\n✓ Download finished successfully!');
    return 0;
  } catch (error) {
    console.error(`\n✗ Download failed: ${error.message}`);
    return 1;
  } finally {
    rl.close();
  }
}

main().then(process.exit);

File: ytmp3.py
#!/usr/bin/env python3
"""
ytmp3 - A simple CLI tool to download YouTube playlists as MP3 files
"""

import os
import sys
import subprocess
from datetime import datetime
import yt_dlp


def get_default_output_dir():
    """Generate default output directory name based on current date"""
    return datetime.now().strftime("%Y-%m-%d-ytmp3")


def download_as_mp3(url, output_path):
    """Download YouTube video/playlist as MP3 using yt-dlp"""
    ydl_opts = {
        'format': 'bestaudio/best',
        'postprocessors': [{
            'key': 'FFmpegExtractAudio',
            'preferredcodec': 'mp3',
            'preferredquality': '192',
        }],
        'outtmpl': os.path.join(output_path, '%(title)s.%(ext)s'),
    }

    try:
        with yt_dlp.YoutubeDL(ydl_opts) as ydl:
            ydl.download([url])
        print(f"\nDownload completed! Files saved to: {output_path}")
    except Exception as e:
        print(f"Error downloading: {e}")
        return False

    return True


def main():
    print("=== ytMP3 - YouTube to MP3 Downloader ===\n")

    # Get YouTube URL
    url = input("Enter the YouTube video or playlist URL: ").strip()
    if not url:
        print("Error: URL cannot be empty")
        return 1

    # Get custom output directory
    default_dir = get_default_output_dir()
    custom_dir = input(f"Enter download location (default: {default_dir}): ").strip()

    if not custom_dir:
        output_path = default_dir
    else:
        output_path = custom_dir

    # Create output directory if it doesn't exist
    if not os.path.exists(output_path):
        try:
            os.makedirs(output_path)
            print(f"Created directory: {output_path}")
        except Exception as e:
            print(f"Error creating directory: {e}")
            return 1

    # Confirm before starting
    input("\nPress Enter to start downloading...")

    # Start download
    print("\nStarting download...")
    success = download_as_mp3(url, output_path)

    if success:
        print("\n✓ Download finished successfully!")
    else:
        print("\n✗ Download failed!")
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main())